lib/spack/spack/database.py

```diff
@@ -0,0 +1,290 @@
+##############################################################################
+# Copyright (c) 2013, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory.
+#
+# This file is part of Spack.
+# Written by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://scalability-llnl.github.io/spack
+# Please also see the LICENSE file for our notice and the LGPL.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License (as published by
+# the Free Software Foundation) version 2.1 dated February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program; if not, write to the Free Software Foundation,
+# Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+import os
+import sys
+import inspect
+import glob
+import imp
+
+import time
+import copy
+import errno
+
+from external import yaml
+from external.yaml.error import MarkedYAMLError
+
+import llnl.util.tty as tty
+from llnl.util.filesystem import join_path
+from llnl.util.lang import *
+
+import spack.error
+import spack.spec
+from spack.spec import Spec
+from spack.error import SpackError
+from spack.virtual import ProviderIndex
+from spack.util.naming import mod_to_class, validate_module_name
+
+
+def _autospec(function):
+    """Decorator that automatically converts the argument of a single-arg
+       function to a Spec."""
+    def converter(self, spec_like, **kwargs):
+        if not isinstance(spec_like, spack.spec.Spec):
+            spec_like = spack.spec.Spec(spec_like)
+        return function(self, spec_like, **kwargs)
+    return converter
+
+
+class Database(object):
+    def __init__(self,root,file_name="specDB.yaml"):
```

Comment:
I don't have a preference about underscores before config files.

I don't see there being any issue with grouping files in ls--the 'ls 
*.yaml' command will work fine.  I'm also not particularly worried about 
users editting our internal files.  I also don't see there being any 
drawbacks to adding an underscore before the filenames.

On Wed, 26 Aug 2015, Todd Gamblin wrote:
> 
> In lib/spack/spack/database.py:
> 
> > +from spack.virtual import ProviderIndex
> > +from spack.util.naming import mod_to_class, validate_module_name
> > +
> > +
> > +def _autospec(function):
> > +    """Decorator that automatically converts the argument of a single-arg
> > +       function to a Spec."""
> > +    def converter(self, spec_like, **kwargs):
> > +        if not isinstance(spec_like, spack.spec.Spec):
> > +            spec_like = spack.spec.Spec(spec_like)
> > +        return function(self, spec_like, **kwargs)
> > +    return converter
> > +
> > +
> > +class Database(object):
> > +    def __init__(self,root,file_name="specDB.yaml"):
> 
> I'm debating adopting a convention to name top-level files like this with a
> starting _. That puts the config files in one place when you ls, and it
> marks them as something special and not necessarily user-configurable. For
> package repositories, I went with _repo.yaml. For the install DB, how about
> we call it _index.yaml or something like that? MacPorts calls its equivalent
> of this file PortIndex. @mplegendre: any preferences on this?